src/libraries/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs

```diff
@@ -31,6 +33,90 @@ public static Type GetNullableType(this Type type)
             return type;
         }
 
+        /// <summary>
+        /// This is an alternative to <see cref="GetNullableType" /> that will throw
+        /// if dynamic code is required. Some common primitive types are special-cased.
+        /// </summary>
+        public static Type LiftPrimitiveOrThrow(this Type type)
+        {
+            if (RuntimeFeature.IsDynamicCodeSupported)
+            {
+#pragma warning disable IL3050
+                // Analyzer doesn't yet understand feature switches
+                return GetNullableType(type);
+#pragma warning restore IL3050
+            }
+            if (!type.IsValueType || IsNullableType(type))
+            {
+                return type;
+            }
+            if (type == typeof(bool))
+            {
+                return typeof(bool?);
+            }
+            else if (type == typeof(int))
+            {
+                return typeof(int?);
+            }
+            else if (type == typeof(long))
+            {
+                return typeof(long?);
+            }
+            else if (type == typeof(float))
+            {
+                return typeof(float?);
+            }
+            else if (type == typeof(double))
+            {
+                return typeof(double?);
+            }
+            else if (type == typeof(uint))
+            {
+                return typeof(uint?);
+            }
+            else if (type == typeof(ulong))
+            {
+                return typeof(ulong?);
+            }
+            else if (type == typeof(byte))
+            {
+                return typeof(byte?);
+            }
+            else if (type == typeof(sbyte))
+            {
+                return typeof(sbyte?);
+            }
+            else if (type == typeof(short))
+            {
+                return typeof(short?);
+            }
+            else if (type == typeof(ushort))
+            {
+                return typeof(ushort?);
+            }
+            else if (type == typeof(char))
+            {
+                return typeof(char?);
+            }
+            else if (type == typeof(decimal))
+            {
+                return typeof(decimal?);
+            }
+            else if (type == typeof(DateTime))
+            {
+                return typeof(DateTime?);
+            }
+            else if (type == typeof(DateTimeOffset))
+            {
+                return typeof(DateTimeOffset?);
+            }
+            else if (type == typeof(TimeSpan))
+            {
+                return typeof(TimeSpan?);
+            }
+            throw new NotSupportedException(Strings.LiftingInExpressionRequiresDynamicCode(type));
```

Comment:
There's no way to provide the developer with recourse in this case? Previously if they ensured the type nullable type was available, it would "just work", but now they have no way to make it work, right?

Reply:
Correct. The philosophy is: it's better to deterministically fail than sometimes succeed. Ideally we'd separate the safe code path from the unsafe code path but I don't see a good way of doing that right now.